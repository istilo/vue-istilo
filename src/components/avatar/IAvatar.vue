<script setup lang="ts">
import { computed } from 'vue';

const props = defineProps<{
	/**
	 * The color for the avatar background
	 * @values primary, secondary
	 */
	color?: string;

	/**
	 * A count for group of users that exceeds on the limit
	 */
	count?: number;

	/**
	 * Text for the avatar, atleast 1 or 2 characters mostly used
	 * for user Initials
	 */
	label?: string;

	/**
	 * Specific size or dimension
	 * @values xsmall, small, medium, large, xlarge
	 */
	size?: string;

	/**
	 * Image source of path
	 */
	src?: string;
}>();

const getClasses = computed(() => {
	return [
		'i-avatar',
		props.color ? `i-avatar-${props.color}` : '',
		props.size ? `i-avatar-${props.size}` : '',
		props.src ? 'i-avatar-image' : '',
	];
});

const getContent = computed(() => {
	return props.count ? (props.count > 9 ? '+9' : `+${props.count}`) : props.label;
});
</script>

<template>
	<div :class="getClasses">
		<img
			v-if="props.src"
			:src="props.src"
			:alt="`${props.label} image`" />

		<span v-else>
			{{ getContent }}
		</span>
	</div>
</template>
