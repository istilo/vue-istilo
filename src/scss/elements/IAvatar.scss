@import '@/scss/helpers';
@import '@/scss/functions';
@import '@/scss/mixins';

// ============================================================
// Name:            Avatar
// Description:     Component to create placeholder or wrapper of the user
//                  image
//
// Component:       `#{$prefix}avatar`
//
// Modifiers:       `#{$prefix}avatar-image`
//                  `#{$prefix}avatar-primary`
//                  `#{$prefix}avatar-xsmall`
//                  `#{$prefix}avatar-small`
//                  `#{$prefix}avatar-medium`
//                  `#{$prefix}avatar-large`
//                  `#{$prefix}avatar-xlarge`
//
// ============================================================

// Variables
// ============================================================

$avatar-border-radius: 100% !default;

$avatar-height-xs: 24px !default;
$avatar-width-xs: $avatar-height-xs !default;
$avatar-font-size-xs: 10px !default;

$avatar-height-s: 32px !default;
$avatar-width-s: $avatar-height-s !default;
$avatar-font-size-s: 12px !default;

$avatar-height: 48px !default;
$avatar-width: $avatar-height !default;
$avatar-font-size: 16px !default;

$avatar-height-m: 64px !default;
$avatar-width-m: $avatar-height-m !default;
$avatar-font-size-m: 20px !default;

$avatar-height-l: 96px !default;
$avatar-width-l: $avatar-height-l !default;
$avatar-font-size-l: 36px !default;

$avatar-height-xl: 128px !default;
$avatar-width-xl: $avatar-height-xl !default;
$avatar-font-size-xl: 48px !default;

$avatar-background: background-color(wrapper) !default;
$avatar-color: color(default) !default;

$avatar-primary-background: theme-color(primary) !default;
$avatar-primary-color: color-yiq($avatar-primary-background) !default;

$avatar-secondary-background: theme-color(secondary) !default;
$avatar-secondary-color: color-yiq($avatar-secondary-background) !default;

$avatar-margin-right: margin(xsmall) !default;

/* ============================================================
	Component: Avatar
   ============================================================ */

/*
 * 1. Set dimension
 * 2. Set typography style
 * 3. Make the text horizontal center
 * 4. Set background color
 */
.#{$prefix}avatar {
	display: inline-block;

	/* 1 */
	height: $avatar-height;
	width: $avatar-width;

	/* 3 */
	font-size: $avatar-font-size;
	line-height: $avatar-height;
	color: $avatar-color;

	/* 3 */
	text-align: center;

	/* 4 */
	background-color: $avatar-background;
	border-radius: $avatar-border-radius;

	overflow: hidden;
}

/*
 * Add margin
 */

.#{$prefix}avatar + * {
	margin-right: $avatar-margin-right;
}

/* Image modifiers
   ============================================================ */

.#{$prefix}avatar.#{$prefix}avatar-image {
	line-height: 0;
}

.#{$prefix}avatar.#{$prefix}avatar-image > img {
	height: 100%;
	width: 100%;
	object-fit: cover;
	object-position: center;
}

/* Size modifiers
   ============================================================ */

.#{$prefix}avatar-xsmall {
	height: $avatar-height-xs;
	width: $avatar-width-xs;
	font-size: $avatar-font-size-xs;
	line-height: $avatar-height-xs;
}

.#{$prefix}avatar-small {
	height: $avatar-height-s;
	width: $avatar-width-s;
	font-size: $avatar-font-size-s;
	line-height: $avatar-height-s;
}

.#{$prefix}avatar-medium {
	height: $avatar-height-m;
	width: $avatar-width-m;
	font-size: $avatar-font-size-m;
	line-height: $avatar-height-m;
}

.#{$prefix}avatar-large {
	height: $avatar-height-l;
	width: $avatar-width-l;
	font-size: $avatar-font-size-l;
	line-height: $avatar-height-l;
}

.#{$prefix}avatar-xlarge {
	height: $avatar-height-xl;
	width: $avatar-width-xl;
	font-size: $avatar-font-size-xl;
	line-height: $avatar-height-xl;
}

/* Color modifiers
   ============================================================ */

.#{$prefix}avatar-primary {
	background-color: $avatar-primary-background;
	color: $avatar-primary-color;
}

.#{$prefix}avatar-secondary {
	background-color: $avatar-secondary-background;
	color: $avatar-secondary-color;
}
