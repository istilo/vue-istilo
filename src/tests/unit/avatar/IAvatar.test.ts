import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { shallowMount, VueWrapper, enableAutoUnmount } from '@vue/test-utils';

import IAvatar from '@/components/avatar/IAvatar.vue';

enableAutoUnmount(afterEach);

describe('Istilo Avatar', () => {
	let wrapper: VueWrapper<any>;

	beforeEach(() => {
		wrapper = shallowMount(IAvatar, {
			propsData: {
				label: 'AV',
			},
		});

		expect(wrapper.exists()).toBe(true);
	});
	afterEach(() => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('Should set a default class and label', () => {
		expect(wrapper.classes()).toContain('i-avatar');
		expect(wrapper.find('span').text()).toBe('AV');
	});

	it('Should set a color modifier class name', async () => {
		await wrapper.setProps({
			color: 'primary',
		});

		expect(wrapper.classes()).toContain('i-avatar-primary');
	});

	it('Should render a count', async () => {
		await wrapper.setProps({
			count: 3,
		});

		expect(wrapper.find('span').text()).toBe('+3');
	});

	it('Should limit the count more than 9', async () => {
		await wrapper.setProps({
			count: 12,
		});

		expect(wrapper.find('span').text()).toBe('+9');
	});

	it('Should set a size modifier class name', async () => {
		await wrapper.setProps({
			size: 'xlarge',
		});

		expect(wrapper.classes()).toContain('i-avatar-xlarge');
	});

	it('Should render the image source that has been set', async () => {
		const imageSrc = 'https://statinfer.com/wp-content/uploads/dummy-user.png';
		await wrapper.setProps({
			src: imageSrc,
		});

		expect(wrapper.find('img')).toBeDefined();
		expect(wrapper.find('img').attributes('src')).toBe(imageSrc);
		expect(wrapper.find('img').attributes('alt')).toBe('AV image');
	});
});
