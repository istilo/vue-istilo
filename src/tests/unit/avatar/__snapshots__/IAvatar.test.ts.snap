// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Istilo Avatar > Should limit the count more than 9 1`] = `"<div class=\\"i-avatar\\"><span>+9</span></div>"`;

exports[`Istilo Avatar > Should render a count 1`] = `"<div class=\\"i-avatar\\"><span>+3</span></div>"`;

exports[`Istilo Avatar > Should render the image source that has been set 1`] = `"<div class=\\"i-avatar i-avatar-image\\"><img src=\\"https://statinfer.com/wp-content/uploads/dummy-user.png\\" alt=\\"AV image\\"></div>"`;

exports[`Istilo Avatar > Should set a color modifier class name 1`] = `"<div class=\\"i-avatar i-avatar-primary\\"><span>AV</span></div>"`;

exports[`Istilo Avatar > Should set a default class and label 1`] = `"<div class=\\"i-avatar\\"><span>AV</span></div>"`;

exports[`Istilo Avatar > Should set a size modifier class name 1`] = `"<div class=\\"i-avatar i-avatar-xlarge\\"><span>AV</span></div>"`;
